sql/core/src/main/scala/org/apache/spark/sql/execution/adaptive/OptimizeSkewedPartitions.scala

```diff
@@ -0,0 +1,281 @@
+            estimateMapIdStartIndices(right, partitionId, rightMedSize)
+          } else {
+            Array(0)
+          }
+
+          for (i <- 0 until leftMapIdStartIndices.length;
+               j <- 0 until rightMapIdStartIndices.length) {
+            val leftEndMapId = if (i == leftMapIdStartIndices.length - 1) {
+              getShuffleStage(left).plan.shuffleDependency.rdd.partitions.length
+            } else {
+              leftMapIdStartIndices(i + 1)
+            }
+            val rightEndMapId = if (j == rightMapIdStartIndices.length - 1) {
+              getShuffleStage(right).
+                plan.shuffleDependency.rdd.partitions.length
+            } else {
+              rightMapIdStartIndices(j + 1)
+            }
+            // For the skewed partition, we set the id of shuffle query stage to -1.
+            // And skip this shuffle query stage optimization in 'ReduceNumShufflePartitions' rule.
+            val leftSkewedReader =
+              PostShufflePartitionReader(getShuffleStage(left).copy(id = -1),
+                partitionId, leftMapIdStartIndices(i), leftEndMapId)
+
+            val rightSkewedReader =
+              PostShufflePartitionReader(getShuffleStage(right).copy(id = -1),
+                partitionId, rightMapIdStartIndices(j), rightEndMapId)
+
+            subJoins +=
+              SortMergeJoinExec(leftKeys, rightKeys, joinType, condition,
+                leftSkewedReader, rightSkewedReader)
+          }
+        }
+      }
+      logInfo(s"skewed partition number is ${skewedPartitions.size}")
+      if (skewedPartitions.size > 0) {
+        getShuffleStage(left).skewedPartitions = skewedPartitions
+        getShuffleStage(right).skewedPartitions = skewedPartitions
+        UnionExec(subJoins.toList)
+      } else {
+        smj
+      }
+  }
+
+  override def apply(plan: SparkPlan): SparkPlan = {
+    if (!conf.adaptiveSkewedJoinEnabled) {
+      return  plan
+    }
+
+    def collectShuffleStages(plan: SparkPlan): Seq[ShuffleQueryStageExec] = plan match {
+      case _: LocalShuffleReaderExec => Nil
+      case _: PostShufflePartitionReader => Nil
+      case stage: ShuffleQueryStageExec => Seq(stage)
+      case ReusedQueryStageExec(_, stage: ShuffleQueryStageExec, _) => Seq(stage)
+      case _ => plan.children.flatMap(collectShuffleStages)
+    }
+
+    val shuffleStages = collectShuffleStages(plan)
+
+    if (shuffleStages.length == 2) {
+      // Currently we only support handling skewed join for 2 table join.
+      handleSkewJoin(plan)
+    } else {
+      plan
+
+    }
+  }
+}
+
+case class PostShufflePartitionReader(
+    child: QueryStageExec,
+    partitionIndex: Int,
+    startMapId: Int,
+    endMapId: Int) extends UnaryExecNode {
+
+  override def output: Seq[Attribute] = child.output
+
+  override def doCanonicalize(): SparkPlan = child.canonicalized
+
+  override def outputPartitioning: Partitioning = {
+    UnknownPartitioning(1)
```

Comment:
shall we return `SinglePartition`? otherwise we may add shuffle between `PostShufflePartitionReader` and SMJ

Reply:
Here we may not execute `ensureRequirements` again. So how add shuffle between `PostShufflePartitionReader` and SMJ?

Reply:
ah i see, then there is no difference.